### DISCUSSION TITLE
compiling with visual studio

### DISCUSSION THREAD

**Discussion by exxosuk** [moonglow/FlashForge_Marlin#58] [Q&A]
I have set FF DREAMER for the board, and it compiles OK. But none of the bin files match the ones I compile at all.  So I dare not flash it to try.. 

But is it supposed to say Dremel ?!

![Capture](https://user-images.githubusercontent.com/22227683/132930025-fe51290a-4bb1-402a-886b-31d0010c4df4.JPG)

**Comment by moonglow** [comment]
For my autobuilds under Linux i use shell script `marlin_builder.sh` because i need to release many other configuration, but in your case you need to find and change one line from `Configuration.h`:

Original:
```c
#pragma once
/* you must be sure what you know what you do :) */
#define FF_FLASHAIR_FIX
#define FF_RUSSIAN_FIX
#define FF_M907_PROTECTION
//#define FF_INVENTOR_MACHINE
//#define FF_DREAMER_MACHINE
//#define FF_DREAMER_NX_MACHINE
//#define FF_DREMEL_3D20_MACHINE
//#define FF_EXTRUDER_SWAP
//#define USE_OLD_MARLIN_UI
//#define USE_MKS_UI
```

Dremel + Color UI:
```c
#pragma once
/* you must be sure what you know what you do :) */
#define FF_FLASHAIR_FIX
#define FF_RUSSIAN_FIX
#define FF_M907_PROTECTION
//#define FF_INVENTOR_MACHINE
//#define FF_DREAMER_MACHINE
//#define FF_DREAMER_NX_MACHINE
#define FF_DREMEL_3D20_MACHINE  /* <------------------- HERE */
//#define FF_EXTRUDER_SWAP
//#define USE_OLD_MARLIN_UI
//#define USE_MKS_UI
```

To rise temp you need to change this line:
```c
#define HEATER_0_MAXTEMP 280
```

By default it sets to 280 ( but we have overshoot settings sets to 15, so actual temp be 265 ).

If firmware was build correct you need to encrypt it with `flashforge_firmware_tool` you must build binaries or if you can't you can ask me about it.... 
Firmware encryption ( example for Dremel firmware ):
```
ff_fw_tool -k flashforge123456 -e -i firmware.bin -o encrypted_firmware.bin
```

Now you can update your printer with `encrypted_firmware.bin`

**Comment by exxosuk** [comment]
Thanks for your reply, I understand what you said above. I did not know the firmware had to be encrypted.

I can compile with VS just by clicking the build button, but I am not sure the settings are correct.

Compiling from the command line was a total pain. Firstly I had to figure out how to install GIT. Then running your bash script it complained phyton3 was not installed (python was in vs extensions) after several hours I managed to get it installed but then the script complains platformio is not installed ( when it is because it works when I just click the build button). In the end I gave up because it is like most other software these days, have problems where nobody knows how to solve them, and everything just seems to be a hack or a workaround to fix it.

I even spent some days trying to compile with the Arduino compiler, but it is somewhat unclear which board settings I should be using in the software, even so, even after selecting the printer in configuration.h it refuses to compile and complains there is no printer selected.

Editing the settings in the configuration files and such is simply enough to do and I understand that part. But it still thinks Dremel is setup..

```c
/* you must be sure what you know what you do :) */
#define FF_FLASHAIR_FIX
#define FF_RUSSIAN_FIX
#define FF_M907_PROTECTION
//#define FF_INVENTOR_MACHINE
#define FF_DREAMER_MACHINE
//#define FF_DREAMER_NX_MACHINE
//#define FF_DREMEL_3D20_MACHINE
//#define FF_EXTRUDER_SWAP
//#define USE_OLD_MARLIN_UI
//#define USE_MKS_UI
```

Hence the image in my first post, so I do not think this is correct. It is like VS is totally ignoring the configuration file. But if I rename "Dremel" to "Dremel2"  shows this new name in VS. so it is processing the configuration file. 

```c
// Name displayed in the LCD "Ready" message and Info menu
#if ENABLED( FF_DREMEL_3D20_MACHINE )
#define CUSTOM_MACHINE_NAME "Dremel2"
#else
#define CUSTOM_MACHINE_NAME "FlashForge"
#endif
```

But then it is totally ignoring the dreamer selection. So I am confused how I seem to correctly set the dreamer printer, and yet it totally ignores it and thinks the Dremel is selected still.

**Comment by lshachar** [comment]
Hi @exxosuk, your last reply is from just over two years ago, so I hardly think it's relevant for you, however I just had to work through (and solve) exactly your problem, and so I'm sure the solution can help someone else too.

First, let me assure you that you're fine with uploading any .bin file you'd like, even if you're not sure if it will work. That is because FF Firmware Writer does not change the bootloader area of the printer's memory. So even if the firmware is corrupted or wrong or even empty, the printer would be unusable or stuck in a boot loop, but you can just go right ahead and flash a different firmware / original fw, and then it will work fine again.

Second, I found out that the 'Dremel' machine name in auto build marlin that you're referring to in your original post - that's just a bug and you shouldn't worry about that. How did I find that out? I changed the last block of code that you've attached (in configuration.h) to:
```
// Name displayed in the LCD "Ready" message and Info menu
#if (0)     // always off, so the block won't get compiled
#define CUSTOM_MACHINE_NAME "Dremel_check"
#else
#define CUSTOM_MACHINE_NAME "FlashForge"
#endif
```

By just saving the file you could already see that auto build marlin in platformIO ide, picks up on the custom machine name and displays it (Dremel_check).
That means there's a bug in auto build marlin and it's displaying 'Dremel' even when that code definitely won't get compiled! So let's move on

You can either use the color UI or the old marlin UI. Using the MKS ui takes some hardware modifications. so make sure this block is commented out. to use the color ui:
```
/* Select UI type */ 
//#define USE_OLD_MARLIN_UI		// to use color UI, leave both of these UI types undefined.
//#define USE_MKS_UI
```

I'll write for completeness sake that you need to uncomment your printer here:
```
/* Select one of the printers below */
//#define FF_INVENTOR_MACHINE
#define FF_DREAMER_MACHINE
//#define FF_DREAMER_NX_MACHINE
//#define FF_DREMEL_3D20_MACHINE
//#undef FF_DREMEL_3D20_MACHINE
```
(I know you have done that)
uncomment black or silver pulley according to your printer, and you're finally done with the basic setup.
Now you can use auto build marlin to create your `firmware.bin` file.
Then you need to encrypt the firmware. If you're on windows like myself, get this [precompiled windows binary encryption/decryption tool](https://github.com/moonglow/FlashForge_Marlin/discussions/139)

There's an `encrypt.bat` file there, but the password in it is for a dremel. So for all other printers, edit the bat file and put the other password in, as described in several places such as [here](https://github.com/moonglow/flashforge_firmware_tool).

now that your firmware is encrypted, finally use FF Firmware writer to upload the firmware to the printer. [directions and tool for windows are here](https://github.com/moonglow/FlashForge_Marlin/wiki/Windows-instalation-guide)

That's it! Now you're finally up and running :)

**Comment by exxosuk** [comment]
I gave up in the end and replaced the board with something more user-friendly and proper marlin support. But I binned the printer not long afterwards. Was just a neverending nightmare of stuff wrong with it. I had pretty much rebuilt the thing in the end. Wish I had never bought the thing.
